lib/Rules/PasswordHistory.php

```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ *
+ * @author Semih Serhat Karakaya <karakayasemi@itu.edu.tr>
+ * @copyright Copyright (c) 2018, ownCloud GmbH
+ * @license GPL-2.0
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\PasswordPolicy\Rules;
+
+use OCA\PasswordPolicy\Db\OldPasswordMapper;
+use OCP\IL10N;
+use OCP\IUserSession;
+
+class PasswordHistory extends Base {
+
+	/** @var IUserSession */
+	protected $userSession;
+
+	/** @var OldPasswordMapper */
+	protected $mapper;
+
+	/**
+	 * @param IL10N $l10n
+	 */
+	public function __construct(IL10N $l10n, IUserSession $userSession, OldPasswordMapper $mapper) {
+		parent::__construct($l10n);
+		$this->userSession = $userSession;
+		$this->mapper = $mapper;
+	}
+
+	public function verify($password, $val) {
+		$oldPasswords = $this->mapper->getOldPasswords(
+			$this->userSession->getUser()->getUID(),

```

Comment:
getUser can return null

Reply:
You are right but have you any suggestion for that?  'OCP\User::validatePassword' and 'OCP\Share::validatePassword' events are triggering this function and they are only supplying a password as the event argument. How can I learn password belongs to which user? We may need to add the user to the event arguments in the core.

Reply:
> We may need to add the user to the event arguments in the core.

should be the case already - if not the event was design the wrong way

E.g. the password have be changed via the terminal - there is no user session in this case.

Please adopt the event in core if necessary.

Thanks for pointing this out!

Reply:
as far as I can see, we have 4 validatePassword event emitting calls in the core and it is not possible to add the user object to event arguments in the all of them. 

https://github.com/owncloud/core/blob/92b658a60e6c28ef9dd7594384180e0d73e96660/lib/private/User/User.php#L253 : we can add here, also maybe only adding here can be sufficient for our purpose.

https://github.com/owncloud/core/blob/92b658a60e6c28ef9dd7594384180e0d73e96660/lib/private/User/Manager.php#L333 since there is no user yet, we can add the only uid in here, but since it user creation event, we don't need to apply password history to the newly created user.

Also, in sharing case, should we apply password history rule? If yes, we should use the creator of share in the validatePassword event.
https://github.com/owncloud/core/blob/92b658a60e6c28ef9dd7594384180e0d73e96660/lib/private/Share20/Manager.php#L174 
https://github.com/owncloud/core/blob/92b658a60e6c28ef9dd7594384180e0d73e96660/lib/private/Share/Share.php#L2867

@DeepDiver1975 what are your thoughts? IMHO, to ensure compatibility with all events, somehow we can add the only uid string in the validatePassword events. If you agree that, I can create a PR in the core for that.

Reply:
good points ... once again.

Yes - adding the uid should be fine. For the share password we dont set the uid because it is not reasonable. If no uid is set it is also pointless to check for the password history.